### ISSUE TITLE
madx sequence file as input to lattice codes

### ISSUE THREAD

**Issue by moellep** [radiasoft/sirepo#1861]
Requested by a zgoubi user, use the madx sequence file as the common format across zgoubi, elegant and synergia. Synergia already uses that format. Allow interoperability across sirepo codes via import/export as madx sequence file.

**Comment by bruhwiler** [comment]
I would like to generalize this issue.
The work can be charged to the MCOOL 2a project.

For every supported tracking code (elegant, Synergia, Zgoubi and Opal), they should accept a MAD-X sequence file, in addition to their native lattice file format.

Also, every tracking code should provide the option to export a MAD-X lattice file, in addition to the native lattice file format.

**Comment by bnash** [comment]
I really support this idea.  

At the moment, I could really use a translator from Zgoubi to Mad-X so that I can then use that lattice with BMAD.

**Comment by bnash** [comment]
@moellep, the tools for converting from MAD and MAD-X to Zgoubi by V. Ranjbar and F. Lin are [here in the zgoubi repository](https://github.com/radiasoft/zgoubi/tree/master/lattice_translation).

**Comment by moellep** [comment]
MAD-X input/output is now implement of OPAL. The zgoubi MAD-X input/output task is still open.

**Comment by bruhwiler** [comment]
This might be a duplicate bug report, but I can't find the original.
Export of MAD-X lattice file is made available in the UI on production, but attempts to use it result in a server error.
I see the same behavior with elegant and OPAL:
![image](https://user-images.githubusercontent.com/3825435/90053979-70925780-dc98-11ea-8660-22b3fa91090d.png)

Using latest version of Firefox on Win10.

**Comment by bruhwiler** [comment]
Testing results (today) for the import/export of MAD-X lattice files on production:
- [x] export from elegant
- [x] import into elegant
- [x] export from OPAL
- [x] import into OPAL
- FAIL: cannot export from MAD-X
- [x] import into MAD-X
- FAIL: can neither export from nor import into Synergia

**Comment by cchall** [comment]
It does look like some unit conversion isn't happening. For instance exporting an elegant lattice with rf cavities leaves frequency and voltage in Hz/V in the MAD-X file, while MAD-X expects MHz/MV. 

The cavity phase also does not seem to be exported.

**Comment by bruhwiler** [comment]
@moellep -- it would be great, if possible, to sort this out in the next week or two.

@cchall -- could you please assist by providing the necessary details?  Please look at OPAL as well.

**Comment by cchall** [comment]
@bruhwiler Sure, I can look over the import/export for OPAL in more detail

We should probably check bend elements carefully as well. The bend length conventions vary between all codes. 
For SBEND elegant, and I believe MAD-X, use arc length while OPAL uses chord length. I'm not sure of the RBEND conventions off-hand, but they are also probably different.

We may also want to consider adding a default hard-edge field map when importing a MAD-X lattice into OPAL. This is used in the OPAL documentation (http://amas.web.psi.ch/opal/Documentation/2.4/index.html#sec.benchmarks.hard-edge-dipole-comparison-with-elegant) and I've also found it essential for getting reasonable matching between elegant in OPAL when doing particle tracking.

**Comment by bruhwiler** [comment]
Input from an external user --

1) Are there unit tests for the MAD-X conversion feature? 

2) It could be useful to compare pyIOTA, https://github.com/nikitakuklev/pyIOTA
and also Synergia, https://bitbucket.org/fnalacceleratormodeling/synergia2/src/devel3/src/synergia/lattice/madx_adaptors.cc
with Sirepo.

3) These conversion bugs (k3 vs k3l, half vs full gap) are difficult to resolve, and I am still not confident I solved them all.

**Comment by nikitakuklev** [comment]
To clarify, point 3 @bruhwiler had is referring to pyIOTA code, which indeed has little testing.

A couple other potential pitfalls I recall:
- Elegant TILT and ETILT conventions were weird, changed recently (https://www3.aps.anl.gov/forums/elegant/viewtopic.php?p=4246#p4246)
- Edge models - MADX implements usual [linear](https://mad.web.cern.ch/mad/webguide/manual.html#Ch10.S4) one, but elegant has more [exotic options](https://ops.aps.anl.gov/manuals/elegant_latest/elegantsu120.html#x131-13000010.16), which are more appropriate for thick tracking

**Comment by moellep** [comment]
I verified the SBEND distance seems to be consistent between elegant/madx/opal/synergia, at least according to the total distance on the twiss reports. The OPAL RBEND doesn't agree and needs to be looked at. sirepo/OPAL also has field unit problems documented in #2748 which need to be addressed.

Many conversion are missing, including the RFCA units. I'll create a new issue with a list of the ones I can identify. The madx_adaptors.cc file is very helpful for comparison.

We have a simple unit test for the conversion which takes an elegant lattice, exports it as mad-x, and then imports the new file back into elegant and creates a new elegant lattice file. https://github.com/radiasoft/sirepo/blob/master/tests/template/madx_converter_test.py

**Comment by cchall** [comment]
> I verified the SBEND distance seems to be consistent between elegant/madx/opal/synergia, at least according to the total distance on the twiss reports.

@moellep Was this a test of Length conversion for SBENDs? I'm fairly certain the definition OPAL uses is different from elegant.

**Comment by moellep** [comment]
@cchall you are correct the OPAL sbend uses the chord length, elegant uses the arc length. I was confused because I thought the s value on the OPAL beam stat file was the particle path distance, but it is only the sum of the element lengths. OPAL's opal_ElementPositions.txt file does seem to show the actual path positions.

**Comment by cchall** [comment]
@moellep the comparison of OPAL against any other code is particularly tricky since it is such a different approach. I'm glad the OPAL manual can be trusted on this at least!

I also put together a single particle tracking test. After verifying individual parameter definitions I think getting sufficiently close results from particle tracking (or Twiss propagation) is the ultimate test for verifying the conversion. Currently in rsbeams: https://github.com/radiasoft/rsbeams/tree/tracking_comparison/examples/tracking_comparisons/sbend_opal_elegant

It also shows that you need to use the chord length for an SBEND in OPAL to get elegant and OPAL to match.

**Comment by robnagler** [comment]
This is still current.
